keps/sig-scheduling/3370-native-podgroup-api/README.md

```diff
@@ -0,0 +1,1076 @@
+
+```go
+// group scheduling.k8s.io
+type PodGroup struct {
+	Spec   PodGroupSpec
+	Status PodGroupStatus
+}
+
+// PodGroupSpec is a description of a PodGroupSpec.
+type PodGroupSpec struct {
+	// Subsets consist of various kinds of pod sets.
+	// A PodGroup is schedulable if all subsets can be scheduled.
+	Subsets []Subset
+
+	// ScheduleTimeoutSeconds defines the timeout threshold to abort
+	// an in-progress PodGroup-level scheduling attempt.
+	ScheduleTimeoutSeconds *int32
+}
+
+// Subset represents a collection of pods with the same role.
+type Subset struct {
+	// Name is used to distinguish pods with different roles in the
+	// same pod group. Optional if there is only one subset in the 
+	// pod group.
+	Name string
+
+	// MinMember specifies the minimum number of pods required for a 
+	// subset to be operational.
+	MinMember int32
+}
+```
+
+In order to associate a pod with its target PodGroup, we need to add a field call `PodGroup` in the podSpec to declare which PodGroup and Subset the pod belongs to.
+
+```go
+// PodSpec is a description of a pod.
+type PodSpec struct {
+	...
+	PodGroup PodGroupRef
+	...
+}
+
+// PodGroupRef contains enough information to let you identify an owning
+// pod group.
+// - pod.spec.podGroup.name = foo
+// - pod.spec.podGroup.subset = (bar/baz)
+type PodGroupRef struct {
+  // Name is the name of PodGroup.
+  Name   string
+  // Subset is the name of PodGroup.
+  Subset string
+}
+```
+
+In this design, a PodGroupStatus is needed to record the latest status. This
+enables users to quickly know what’s going on with this PodGroup; also some
+integrators can rely on it to make pending PodGroup schedulable. For this, we 
+need to add a new PodGroup Controller in controller-manager.
+
+```go
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+	// Represents the latest available observations of 
+	// a pod group's current state.
+	Conditions []PodGroupCondition
+
+	// Subsets is a list of SubsetStatus which represents 
+	// the current state of a single subset.
+	Subsets []SubsetStatus
+}
+
+// PodGroupCondition describes the state of a pod group 
+// at a certain point.
+type PodGroupCondition struct {
+	// Type of deployment condition.
+	Type PodGroupConditionType
+
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus
+
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time
+
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time
+	
+	// The reason for the condition's last transition.
+	Reason string
+	
+	// A human readable message indicating details about 
+	// the transition.
+	Message string
+}
+
+type PodGroupConditionType string
+
+const (
+	// PodGroupScheduled represents status of the scheduling 
+	// process for this pod group.
+	PodGroupScheduled PodGroupConditionType = "PodGroupScheduled"
```

Comment:
`PodGroupScheduled` represents the pod group is scheduled, not only means it's scheduled successfully. 

> And whether Scheduled enough rather than PodGroupScheduled since it's the status of PodGroupCondition.

`Scheduled` or `PodGroupScheduled` are both ok to me.